WalletWasabi/Services/TransactionCreator.cs

```diff
@@ -0,0 +1,253 @@
+			builder.AddCoins(allowedSmartCoinInputs.Select(c => c.GetCoin()));
+
+			for (int i = 0; i < toSend.Length; i++) // clone
+			{
+				var destination = toSend[i].Script;
+				var amount = toSend[i].Amount;
+				var label = toSend[i].Label;
+				if (amount == Money.Zero)
+				{
+					builder.SendAll(destination);
+				}
+				else
+				{
+					builder.Send(destination, amount);
+				}
+
+				if (subtractFeeFromAmountIndex is int inputIndex && inputIndex == i)
+				{
+					builder.SubtractFees();	
+				}
+			}
+
+			HdPubKey changeHdPubKey = null;
+			if (customChange is Script)
+			{
+				builder.SetChange(customChange);
+			}
+			else
+			{
+				KeyManager.AssertCleanKeysIndexed(isInternal: true);
+				KeyManager.AssertLockedInternalKeysIndexed(14);
+				changeHdPubKey = KeyManager.GetKeys(KeyState.Clean, true).RandomElement();
+				builder.SetChange(changeHdPubKey.P2wpkhScript);
+			}
+
+			builder.SendEstimatedFees(feePerBytes);
+			var psbt = builder.BuildPSBT(false);
+
+			string changeLabel = null;
+			var spentCoins = psbt.Inputs.Select(txin => smartCoinsByOutpoint[txin.PrevOut]).ToArray();
+
+			var realToSend = toSend.Select(t => (label: t.Label, 
+												script: t.Script, 
+												amount: psbt.Outputs.FirstOrDefault(o => o.ScriptPubKey == t.Script)?.Value))
+									.Where(i => i.amount != null)
+									.ToArray();
+
+			if (changeHdPubKey is HdPubKey)
+			{
+				var sb = new StringBuilder();
+				foreach (var item in realToSend)
+				{
+					var corrected = Guard.Correct(item.label);
+					sb.Append($"{corrected}, ");
+				}
+				changeLabel = sb.ToString().TrimEnd(',', ' ');
+				changeHdPubKey.SetLabel(changeLabel, KeyManager);
+			}
+
+			if (!psbt.TryGetFee(out var fee))
+			{
+				throw new InvalidOperationException($"Impossible to get the fees of the PSBT, this should never happen.");
+			}
+			Logger.LogInfo<WalletService>($"Fee: {fee.Satoshi} Satoshi.");
+
+			var vSize = builder.EstimateSize(psbt.GetOriginalTransaction(), true);
+			Logger.LogInfo<WalletService>($"Estimated tx size: {vSize} vbytes.");
+
+			// 6. Do some checks
+			Money totalOutgoingAmountNoFee = realToSend.Select(x => x.amount).Sum();
+			Money totalOutgoingAmount = totalOutgoingAmountNoFee + fee;
+			decimal feePc = (100 * fee.ToDecimal(MoneyUnit.BTC)) / totalOutgoingAmountNoFee.ToDecimal(MoneyUnit.BTC);
+
+			if (feePc > 1)
+			{
+				Logger.LogInfo<WalletService>($"The transaction fee is {feePc:0.#}% of your transaction amount.{Environment.NewLine}"
+					+ $"Sending:\t {totalOutgoingAmount.ToString(fplus: false, trimExcessZero: true)} BTC.{Environment.NewLine}"
+					+ $"Fee:\t\t {fee.Satoshi} Satoshi.");
+			}
+			if (feePc > 100)
+			{
+				throw new InvalidOperationException($"The transaction fee is more than twice as much as your transaction amount: {feePc:0.#}%.");
+			}
+
+			if (spentCoins.Any(u => !u.Confirmed))
+			{
+				Logger.LogInfo<WalletService>("Unconfirmed transaction are being spent.");
```

Comment:
Unconfirmed transaction `is` being spent.

Reply:
fixed